Webfonts: register and enqueue fonts by font-family

zaguiini opened the pull request:
## What?

Part of https://github.com/WordPress/gutenberg/issues/39332.

## Why?

It's an attempt to make the webfonts API more flexible and save transferred data. We're now registering font families (individual font faces grouped by font-family)  and enqueueing them as needed.

1. We more closely match the behavior of wp scripts and styles and accommodate functionality that WordPress developers are already familiar with.
2. We allow API consumers more flexibility in how and when they enqueue webfonts, rather than forcing them to enqueue all fonts indiscriminately.

## How?

We're creating a hash of the font-family attribute and grouping registered font faces to their respective font families;

We then introduce enqueueing and do not add registered font-families to the output by default -- they must be explicitly added through enqueueing. All fonts are enqueued when opening the editor so fonts picked when making changes show up correctly.

Webfonts should be enqueued after the `after_setup_theme` hook runs because that's where registration happens. So we should recommend hooking into `init` or `wp_loaded` to enqueue fonts since it's guaranteed to be registered at the time that these hooks are run.

### Testing a webfont loaded via `theme.json`:

Using the Bennett theme as an example, we edit its `theme.json` file and replace `fontFamilies` with the following:
```json

"fontFamilies": [
    {
        "fontFamily": "Roboto",
        "name": "Roboto",
        "slug": "roboto",
        "fontFace": [
            {
                "fontFamily": "Roboto",
                "fontWeight": "900",
                "fontStyle": "normal",
                "fontStretch": "normal",
                "src": [ "file:./assets/fonts/Roboto-Regular.ttf" ]
            },
        ]
    }
],
```
Then, we create a `functions.php` in the `bennett` theme directory and add the following:
```php
<?php

add_action( 'init', function() {
	wp_enqueue_webfont( 'Roboto' );
} );
```

### Testing a webfont programmatically through PHP:

Next, to test the PHP registration using the `wp_register_webfont` function, in `functions.php` add the following:

```php
add_action( 'after_setup_theme', function() {
	wp_register_webfont(
		array(
			'font-family'  => 'Roboto',
			'font-style'   => 'normal',
			'font-stretch' => 'normal',
			"font-weight" => "900",
			'src'          => get_theme_file_uri( '/assets/fonts/Roboto-Regular.ttf' ),
		)
	);
} );

add_action( 'wp_loaded', function() {
	wp_enqueue_webfont( 'Roboto' );
} );
```

Refresh the page and Roboto should be there. You can try adding another font of the same family in italic (`font-style: italic`)  and changing the src. It should load both fonts.

When you remove `wp_enqueue_webfont`, Roboto should not be loaded anymore.

zaguiini committed:
Separate between registering and enqueueing webfonts

zaguiini committed:
Group registered webfonts by font family

zaguiini committed:
Add registered font-family groups to theme.json

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts-provider-local.php:
```diff
@@ -227,9 +227,9 @@ private function compile_src( $font_family, array $value ) {
-			if ( 0 === strpos( $item['url'], get_site_url() ) ) {
-				$item['url'] = wp_make_link_relative( $item['url'] );
+			if ( 0 === strpos( $item['url'], 'file:./' ) ) {
+				$absolute_path_to_url = get_stylesheet_directory_uri() . '/' . str_replace( 'file:./', '', $item['url'] );
```
A couple of things here:

- `get_stylesheet_directory_uri()` limits this functionality to only themes - meaning plugins cannot use the `file:./` notation. Is this intended? Is `file:./` notation now only available for themes, meaning plugins are required to register local web fonts with absolute path to where the file(s) exist in the plugin?

- Micro-optimization: `get_site_url()` and/or `get_stylesheet_directory_uri()` can be cached within the provider object (either through memoization within this method or as a property) and then the value used here with the `foreach` loop.

- `0 === strpos( $item['url'], 'file:./' )` can be replaced with the more readable `str_starts_with()` function that was introduced in WP 5.9.

zaguiini replied:
>Is this intended? [...] meaning plugins are required to register local web fonts with absolute path to where the file(s) exist in the plugin

Yes -- for now. In `theme.json` you cannot specify a dynamic path that'd get resolved relative to the theme folder, that's why this notation exists.

In plugins/PHP realm, you can always pass the full URL built from the helper methods. We might make it work for plugins and themes in the future. This is just moving existing code around, not adding anything there.

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -16,7 +16,15 @@ class WP_Webfonts {
+	/**
+	 * An array of enqueued webfonts.
+	 *
+	 * @access private
```
```suggestion
```
Core doesn't use `@access` to identify properties visibility. This annonation can be removed.

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -81,19 +107,45 @@ public function get_providers() {
-	 * Get the font ID.
+	 * Get the font slug.
 	 *
 	 * @param array $font The font arguments.
```
This method now accepts 2 different data types: `array` or `string`.  
```suggestion
	 * @param array|string $to_convert The value to convert into a slug. Expected as the web font's array or a font-family as a string.
```

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -81,19 +107,45 @@ public function get_providers() {
 	 */
-	public function get_font_id( $font ) {
-		return sanitize_title( "{$font['font-family']}-{$font['font-weight']}-{$font['font-style']}-{$font['provider']}" );
+	public function get_font_slug( $font ) {
+		if ( is_string( $font ) ) {
+			return sanitize_title( $font );
+		}
+
+		return sanitize_title( $font['font-family'] );
 	}
```
Hmm, does this need to be publicly exposed within the API? Seems to be an internal function that assumes it will receive an already validated array or font-family. I'd suggest making it `private` as part of the internals.

I'd also suggest refactoring for readability and reducing the risks of future changes touching only 1 of the `sanitize_title()` instead of both.

```suggestion
	private function get_font_slug( $to_convert ) {
		if ( is_array( $to_convert ) ) {
			$to_convert = $to_convert['font-family'];
		}	

		return sanitize_title( $to_convert );
	}
```

`$to_convert` is more appropriate and descriptive than `$webfont`. Why? The param/var name of`$webfont`'s is (or should be) consistently referring to the web font's configuration (which includes it's font-face properties). Here, the method will receive a webfont's configuration (as an array) or the font-family (as a string). Since it's different kinds of data being passed into the method for conversion into a slug, a more generic name is needed for readability and future contributions.

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -238,22 +290,30 @@ public function generate_and_enqueue_editor_styles() {
 	 *
```
```suggestion
	 *
	 * @param array[] $font_families Font families and each of their webfonts.
```

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -81,19 +107,45 @@ public function get_providers() {
+
+			$this->registered_webfonts[ $slug ][] = $font;
+		}
+	}
```
A few suggestions:
- Type hint the method's parameter with `array` as `array` is the only expected data type.
- For consistency, use `$webfont` instead of `$font` as `webfont` refers to the configuration of a web font (i.e. refers to the list of arguments).
- For readability, convert the `if` conditional into a guard clause and bail out early if not valid. This shifts all the code remaining in the method to the left and clearly shows the intent that if not validate, it's done and no work will happen.

```php
	/**
	 * Register a webfont.
	 *
	 * @param array $webfont The web font's arguments.
	 */
	public function register_font( array $webfont ) {
		$webfont = $this->validate_font( $webfont );

		// If not valid, bail out.
		if ( empty( $webfont ) ) {
			return;
		}

		$slug = $this->get_font_slug( $webfont );

		// Initialize a new font-family collection.
		if ( ! isset( $this->registered_webfonts[ $slug ] ) ) {
			$this->registered_webfonts[ $slug ] = array();
		}

		$this->registered_webfonts[ $slug ][] = $webfont;
	}
```

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -81,19 +107,45 @@ public function get_providers() {
+		$slug = $this->get_font_slug( $font_family );
+
+		if ( isset( $this->enqueued_webfonts[ $slug ] ) ) {
+			_doing_it_wrong( __FUNCTION__, sprintf( __( 'The "%s" font family is already enqueued.' ), $slug ), '6.0.0' );
```
Each case of `__doing_wrong()` is being reviewed and more often than not replaced by `trigger_error()`. Why? Per @azaozz, `_doing_it_wrong()` is "yelling" at the developer and it should be reserved for the most egregious instances.

What about here? Does this qualify as an egregious instance? I don't think "already enqueued" rises to the level of yelling at the developer. Instead, I think using `trigger_error()` is more appropriate as it will log the issue to alert the developer that action is needed.

```suggestion
			trigger_error( 
				sprintf(
					/* translators: %s unique slug to identify the webfont */
					__( 'The "%s" font family is already enqueued.', 'gutenberg' ),
					$slug
				)
			);
```

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -81,19 +107,45 @@ public function get_providers() {
+		if ( ! isset( $this->registered_webfonts[ $slug ] ) ) {
+			_doing_it_wrong( __FUNCTION__, sprintf( __( 'The "%s" font family is not registered.' ), $slug ), '6.0.0' );
+			return false;
 		}
```
What about this instance of `_doing_it_wrong()`?  Does this qualify as an egregious instance? Yes, I think so as this function should never be called if the webfont is not registered with the API. BUT having `trigger_error()` above and `_doing_it_wrong()` here will be confusing and less readable. Contributors will read the code and stop to say "Huh". Not good. I think instead, using `trigger_error()` here is likely better for readability and consistency within this class.

```suggestion
			trigger_error( 
				sprintf(
					/* translators: %s unique slug to identify the webfont */
					__( 'The "%s" font family is not registered.', 'gutenberg' ),
					$slug
				)
			);
```

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -16,7 +16,15 @@ class WP_Webfonts {
+	/**
+	 * An array of enqueued webfonts.
+	 *
+	 * @access private
+	 * @var array
```
```suggestion
	 * @var array[]
```

Core doesn't use @access to identify properties visibility. This annotation can be removed. Also this property is an array of arrays when populated as denoted by `array[]`.

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -56,12 +64,30 @@ public function init() {
+	/**
+	 * Get the list of enqueued fonts.
+	 *
+	 * @return array
```
```suggestion
	 * @return array[]
```

When populated, this returns an array of font-family arrays.

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -56,12 +64,30 @@ public function init() {
+	/**
+	 * Get the list of all fonts.
 	 *
 	 * @return array
```
```php
	 * @return array[]
```

hellofromtonya commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -56,12 +64,30 @@ public function init() {
-	 * Get the list of fonts.
+	 * Get the list of registered fonts.
+	 *
+	 * @return array
```
```suggestion
	 * @return array[]
```
When populated this returns an array of arrays.

zaguiini committed:
Add wp_enqueue_webfont method

zaguiini committed:
Fix wp_register_webfont description

zaguiini committed:
Do not duplicate font families if registering from both theme.json and API

zaguiini commented:
@hellofromtonya this is ready for review and adapted to your comments.

@aristath please take a look when possible 🙏

jeyip committed:
Rename font to webfont where appropriate

creativecoder commented on lib/compat/wordpress-6.0/class-wp-webfonts.php:
```diff
@@ -76,36 +99,92 @@ public function get_providers() {
+	 *
+	 * @param array|string $to_convert The value to convert into a slug. Expected as the web font's array or a font-family as a string.
+	 */
+	public static function get_font_slug( $to_convert ) {
+		if ( is_array( $to_convert ) ) {
+			if ( isset( $to_convert['font-family'] ) ) {
+				$to_convert = $to_convert['font-family'];
+			} elseif ( isset( $to_convert['fontFamily'] ) ) {
+				$to_convert = $to_convert['fontFamily'];
+			} else {
+				_doing_it_wrong( __FUNCTION__, __( 'Could not determine the font family name.' ), '6.0.0' );
+				return false;
+			}
+		}
+
+		return sanitize_title( $to_convert );
```
How will this handle the same font families registered by different providers? (A local provider font from the theme and a Google Font from a plugin, for example)

zaguiini replied:
It will register both, for now.

We're thinking of creating a separate PR so we can handle font face conflicts in a better way. If the font faces are equal, we'll override it so the last provider wins; if not, we'll just add it to the list, so both font faces are registered.

creativecoder reviewed (commented):
The overall approach is looking good and makes sense to me.

I think this is best reviewed in tandem with #39593, as the first real test is if it helps us remove unused font-face declarations to avoid unnecessarily bloating the html.

zaguiini committed:
Add tests

zaguiini committed:
Fix linter offenses

zaguiini committed:
Fire gutenberg_register_webfonts_from_theme_json on init

zaguiini commented on lib/compat/wordpress-6.0/register-webfonts-from-theme-json.php:
```diff
@@ -122,28 +123,28 @@ function gutenberg_add_registered_webfonts_to_theme_json( $data ) {
 }
 
-add_action( 'wp_loaded', 'gutenberg_register_webfonts_from_theme_json' );
+add_action( 'after_setup_theme', 'gutenberg_register_webfonts_from_theme_json' );
```
This is necessary so we can register the webfonts from `theme.json` as soon as possible.

Because themes can enqueue webfonts registered through `theme.json`, and they do that when hooking into `after_setup_theme` inside their `functions.php` file, running this hook on `wp_loaded` means it's too late to register the fonts.

I found that the default priority works just fine -- plugins are registered before themes. But let me know if this should change.

cc @aristath

zaguiini replied:
And now the tests are failing 😕  Any tips on how to fix it?

zaguiini replied:
Well... Thinking more about it.

Maybe we do not need this change.

Registration of fonts from `theme.json` can happen in `wp_loaded` and programmatically registration of fonts can happen in `after_setup_theme`, that's just fine.

We might need to change the documentation on `wp_enqueue_webfont` so that it's clear that manual enqueueing needs to happen after `wp_loaded` at least, so it's guaranteed that the fonts were registered.

zaguiini replied:
Also -- why is `gutenberg_register_webfonts_from_theme_json` hooked into `wp_loaded` and not into `init`, @aristath?

zaguiini commented on lib/compat/wordpress-6.0/register-webfonts-from-theme-json.php:
```diff
@@ -122,28 +123,28 @@ function gutenberg_add_registered_webfonts_to_theme_json( $data ) {
 }
 
-add_action( 'wp_loaded', 'gutenberg_register_webfonts_from_theme_json' );
+add_action( 'init', 'gutenberg_register_webfonts_from_theme_json' );
```
@aristath any specific reason to hook into `wp_loaded` instead of `init`?

`init` runs earlier but after the theme is loaded. Moving to `init` makes enqueueing fonts in `wp_loaded` (and `init`) possible. Any concerns?

aristath replied:
I don't recall exactly why we used `wp_loaded` here... 🤔 
If we change this, we need to check the following:
* Registering fronts both via `theme.json` and PHP (using the `wp_register_webfonts` function)
* Registered webfonts (via theme.json and PHP) appear in the editor's font-family selector
* CSS for registered webfonts (via theme.json and PHP) is present in the editor
* CSS for registered webfonts (via theme.json and PHP) is present ojn the frontend

If all of the above work as expected, then there's no issue switching the hook to `init` 👍

zaguiini replied:
Great 👍

Nothing changed, so we're good to go.

zaguiini committed:
Move webfonts logic from wordpress-6.0 to experimental

The API, although looking really good, is not ready for Core.
So we must move it from wordpress-6.0 to experimental,
and only add it there when it's ready to enter WordPress.

Relocate files to /lib/experimental.

Per the Gutenberg PHP > File structure documentation
https://github.com/WordPress/gutenberg/tree/trunk/lib#file-structure.

hellofromtonya committed:
Update return types and comment for consistency and matching.

hellofromtonya committed:
Use trigger_log instead of error_log().

hellofromtonya committed:
Fix tests with _doing_it_wrong.

Tests for methods that have a `_doing_it_wrong()` require
an `expectedIncorrectUsage` annotation, proper usage of test fixture methods,
and the usage of `__METHOD__` when invoking `_doing_it_wrong()`.

This commit:

- Fixes the test fixtures to use snake_case format and calling of the parent fixtures.
- Refactors `get_font_slug()` tests into data providers (preferred in Core) and separates tests that will trigger `_doing_it_wrong()` to use the annotation.
- Uses `__METHOD__` instead of `__FUNCTION__` in each `_doing_it_wrong()` instance.

hellofromtonya committed:
Add `get_webfonts_by_provider()` method to handle regrouping.

The `WP_Webfonts::generate_styles()` should not be aware of how to
reorganize the webfonts from grouped by font-family to grouped by
provider. Why? This is a separate task from generating styles for
the webfonts.

This commit creates a private method called `get_webfonts_by_provider()`
to encapsulate the know-how for this reorganizational work.

It also includes a micro-optimization that eliminates an extra
`foreach` loop and the building of an unnecessary `$webfonts`
array of arrays.

hellofromtonya commented on lib/experimental/webfonts.php:
```diff
@@ -76,25 +74,68 @@ function wp_register_webfonts( array $webfonts = array() ) {
+ *
+ * @param string[] $webfonts Font families to be enqueued.
+ */
+function wp_enqueue_webfonts( $webfonts ) {
```
Type hint with `array`.

```suggestion
function wp_enqueue_webfonts( array $webfonts ) {
```

hellofromtonya replied:
Resolved in https://github.com/WordPress/gutenberg/pull/39559/commits/27e7da71a053c7e21c91d83935fbb1fd460f376b.

hellofromtonya commented on lib/experimental/webfonts.php:
```diff
@@ -61,8 +61,6 @@ function wp_webfonts() {
- *                        See {@see WP_Webfonts_Registry::register()} for a list of
- *                        supported arguments for each webfont.
  */
 function wp_register_webfonts( array $webfonts = array() ) {
```
The array of webfonts is required, not optional. Remove the default value.
```suggestion
function wp_register_webfonts( array $webfonts ) {
```

hellofromtonya replied:
Resolved in https://github.com/WordPress/gutenberg/pull/39559/commits/27e7da71a053c7e21c91d83935fbb1fd460f376b.

hellofromtonya commented on lib/experimental/register-webfonts-from-theme-json.php:
```diff
@@ -122,28 +123,28 @@ function gutenberg_add_registered_webfonts_to_theme_json( $data ) {
-			'name'       => $family,
-			'slug'       => sanitize_title( $family ),
-			'fontFace'   => $font_face,
+			'fontFamily' => false !== strpos( $family_name, ' ' ) ? "'{$family_name}'" : $family_name,
```
Use `str_contains()` for readability:

```suggestion
			'fontFamily' => str_contains( $family_name, ' ' ) ? "'{$family_name}'" : $family_name,
```

This yields the same result as using `strpos()` but is more readable. See it in action here https://3v4l.org/VCIg4#v8.0.17.

hellofromtonya replied:
Resolved in https://github.com/WordPress/gutenberg/pull/39559/commits/27e7da71a053c7e21c91d83935fbb1fd460f376b.

hellofromtonya commented on lib/experimental/webfonts.php:
```diff
@@ -76,25 +74,68 @@ function wp_register_webfonts( array $webfonts = array() ) {
+ *
+ * @since 6.0.0
+ *
+ * @param string $webfont Font family to be enqueued.
+ */
+function wp_enqueue_webfont( $webfont ) {
+	wp_webfonts()->enqueue_webfont( $webfont );
 }
```
Return the bool flag from `enqueue_webfont()` to provide an indication if the enqueue was successful or not.
And it change the function's parameter name and comment to match the signature of `WP_Webfonts::enqueue_webfont()`, i.e. enqueues by a font family name.

```suggestion
 * @param string $font_family_name The font family name to be enqueued.
 * @return bool True if successfully enqueued, else false.
 */
function wp_enqueue_webfont( $font_family_name ) {
	return wp_webfonts()->enqueue_webfont( $font_family_name );
}
```

hellofromtonya replied:
Resolved in https://github.com/WordPress/gutenberg/pull/39559/commits/27e7da71a053c7e21c91d83935fbb1fd460f376b.

hellofromtonya commented on lib/experimental/webfonts.php:
```diff
@@ -76,25 +74,68 @@ function wp_register_webfonts( array $webfonts = array() ) {
 function wp_register_webfont( array $webfont ) {
-	wp_webfonts()->register_font( $webfont );
+	wp_webfonts()->register_webfont( $webfont );
+}
```
Return the bool from the method as an indication if the webfont was successfully registered or not. And update the DocBlock comment to ensure the method and helper function match.

hellofromtonya replied:
Resolved in https://github.com/WordPress/gutenberg/pull/39559/commits/27e7da71a053c7e21c91d83935fbb1fd460f376b.

hellofromtonya committed:
Updates from code review.

gziolo commented:
This PR needs a rebase after https://github.com/WordPress/gutenberg/pull/39888 which added guard checks for functions and class declarations to prevent fatal errors when the same code gets added to WordPress core.

hellofromtonya commented:
>This PR needs a rebase after https://github.com/WordPress/gutenberg/pull/39888 which added guard checks for functions and class declarations to prevent fatal errors when the same code gets added to WordPress core.

Done 👍

hellofromtonya commented:
Rebased against latest `trunk` to resolve merge conflicts.

hellofromtonya commented on lib/load.php:
```diff
@@ -101,17 +101,17 @@ function gutenberg_is_experiment_enabled( $name ) {
 require __DIR__ . '/compat/wordpress-6.0/class-wp-rest-block-pattern-categories-controller.php';
 require __DIR__ . '/compat/wordpress-6.0/functions.php';
-require __DIR__ . '/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php';
+require __DIR__ . '/compat/wordpress-6.0/class-wp-theme-json-resolver-6-0.php';
```
🤦‍♀️ my error on resync. `require __DIR__ . '/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php';` should remain and is the reason why the tests are failing. Doh. Resolving now.

hellofromtonya replied:
Resolved in https://github.com/WordPress/gutenberg/pull/39559/commits/cc63012ed9d00f8dced590d218d8cdc19721f88c.

hellofromtonya committed:
Restore loading lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php.

zaguiini commented on lib/load.php:
```diff
@@ -101,6 +101,7 @@ function gutenberg_is_experiment_enabled( $name ) {
 require __DIR__ . '/compat/wordpress-6.0/class-wp-rest-block-patterns-controller.php';
 require __DIR__ . '/compat/wordpress-6.0/class-wp-rest-block-pattern-categories-controller.php';
 require __DIR__ . '/compat/wordpress-6.0/functions.php';
+require __DIR__ . '/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php';
```
@hellofromtonya this needs to be `class-wp-theme-json-6-0` because we're modifying this in `/experimental/class-wp-theme-json-gutenberg`

zaguiini replied:
The reason it needs to stay like that is that the version in `compat` is reading the `theme-i18n.json` file.

zaguiini replied:
We're modifying the resolver, not theme json class. Disregard.

hellofromtonya replied:
This PR is modifying the resolver, i.e. `/compat/wordpress-6.0/class-wp-theme-json-resolver-6-0.php`

hellofromtonya reviewed (approved):
Scope of this PR - LGTM 👍 

Other changes (such as tests, wiring theme.json enqueue, etc.) will be done in other PR(s).

scruffian commented on lib/experimental/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -15,54 +15,7 @@
-		return $variations;
-	}
-
+class WP_Theme_JSON_Resolver_Gutenberg extends WP_Theme_JSON_Resolver_6_0 {
```
This file no longer seems to be loaded at all? I'd like to modify `get_theme_data` but I'm not sure how now...

zaguiini replied:
Sorry @scruffian, what do you mean? It is listed in `load.php`: https://github.com/WordPress/gutenberg/blob/trunk/lib/load.php#L111.

scruffian replied:
Sorry, looks like I was stuck in a bad merge state...

bradhogan commented:
Is this now the way fonts should be loaded into a block theme?  Any official documentation on this?

zaguiini commented:
Hey @bradley2083,

Please read https://github.com/WordPress/gutenberg/issues/39332#issuecomment-1099342269.

bradhogan commented:
@zaguiini Thanks for sharing that.  So I'm still not 100% clear but sounds like we still need to use font-face declarations for the time being?

zaguiini commented:
Sorry, @bradley2083.

We now have separated registration and enqueueing of fonts, just like we have with scripts and styles.

https://github.com/WordPress/gutenberg/pull/39988 will enqueue the fonts listed in `theme.json`, even if they're registered programmatically (i.e. through the `wp_register_webfont` method).

After that, we'll work on https://github.com/WordPress/gutenberg/issues/40363, so we can enqueue automatically the fonts picked in global styles.

So no, you won't need to use font-face declarations. The Webfonts API should be sufficient. Right now, you can use it to register and enqueue webfonts through the API.

When https://github.com/WordPress/gutenberg/pull/39988 gets merged, you can use solely `theme.json` to register and enqueue fonts. Note that the picker in global styles won't work for externally/programmatically registered webfonts, only for the fonts explicitly listed in `theme.json`.

This problem will be addressed in https://github.com/WordPress/gutenberg/issues/40363, so externally registered webfonts will be considered when enqueueing fonts to the front-end.

bradhogan commented:
Ok got it.  Thanks @zaguiini I'll start to play around with using `wp_register_webfont` I assume this is too early to have any documentation / examples for this stuff atm?

zaguiini commented:
Yeah, right, @bradley2083. It's not going to be included in 6.0, so too early. But the methods in `webfonts.php` do have some documentation. It should be straightforward to understand what's going on.

gigitux commented:
We noticed that with the new Gutenberg versions (from 13.0) the Global Styles for WC Blocks plugin doesn’t apply anymore.
 
![Screen Capture on 2022-05-03 at 16-08-09](https://user-images.githubusercontent.com/4463174/166484835-a55350e3-a90c-41e8-835c-f50fc80b884f.gif)

It seems that the problem is caused by this PR, in particular, this [this change](https://github.com/WordPress/gutenberg/pull/39559/files#diff-a68821415230540a4e41b902b125c3cc8ed44851c6f5a84817d37160076c51aaR148).

I don't have a lot of context on this part of the code. Could you take a look? Thanks!

cc @zaguiini @hellofromtonya
